### PR TITLE
[autoparallel] add rotor C version

### PR DESCRIPTION
**Pull request by Cypher30**
# What's New?

In this PR I add C extension of rotor dynamic programming table calculation and remove repeated `MetaInfoProp` in rotor solver, right now you should execute the following code before running solver rotor

```python
if META_COMPATIBILITY:
    data_meta = MetaTensor(data, fake_device=next(gm.parameters()).device)
    MetaInfoProp(gm).run(data_meta)
```

As we might use rotor for simplified offload strategy, this modification is crucial as we might run solver over the same graph several time for auto capper heuristic, we don't need to run this MetaInfoProp over and over again. Actually with the new C version dynamic programming, the `MetaInfoProp` itself consumes more time than simply run the dynamic programming.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -191,6 +192,12 @@ def cuda_ext_helper(name, sources, extra_cuda_flags, extra_cxx_flags=[]):
     extra_cxx_flags = ['-std=c++14', '-lcudart', '-lcublas', '-g', '-Wno-reorder', '-fopenmp', '-march=native']
     ext_modules.append(cuda_ext_helper('cpu_adam', ['cpu_adam.cpp'], extra_cuda_flags, extra_cxx_flags))
 
+if build_auto_ckpt_ext:
```
Comment (FrankLeeeee) [inline setup.py:195]: There is no way to specify this parameter by the user.

File: setup.py [hunk 2]
```diff
@@ -191,6 +192,12 @@ def cuda_ext_helper(name, sources, extra_cuda_flags, extra_cxx_flags=[]):
...
+    print(os.path.join(this_dir, "colossalai/fx/passes/algorithms/dynamic_programs.c"))
```
Comment (FrankLeeeee) [inline setup.py:196]: Remove this print statement.

File: setup.py [hunk 3]
```diff
@@ -191,6 +192,12 @@ def cuda_ext_helper(name, sources, extra_cuda_flags, extra_cxx_flags=[]):
...
+    ext_modules.append(
+        Extension("c_version_dp",
```
Comment (FrankLeeeee) [inline setup.py:198]: do not name this way, it is quite meaningless.

Comment (Cypher30) [reply setup.py]: I've fixed those problems, currently we will build the C extension the first time user run the rotor solver, only if C extension building fail will we use the python version of dynamic programming.

**Comment by Cypher30** [comment]
I will build a unit test to compare the solution of python and C version dynamic programs.

**Comment by FrankLeeeee** [comment]
I will convert this PR to draft until it is ready for review again.

**Comment by Cypher30** [comment]
Screen Shot 2022-09-29 at 10 21 58
As the unit test of C version solver consistency is skipped in torch11, the above is the result on torch12.

File: colossalai/fx/passes/algorithms/ckpt_solver_rotor.py [hunk 4]
```diff
@@ -5,9 +5,26 @@
...
+                              stdout=subprocess.PIPE,
+                              stderr=subprocess.PIPE,
+                              shell=True)
+    if result.wait() == 0:
+        print("dynamic_programs_C_version has been built!")
+        from .dynamic_programs_C_version import persistent_compute_table
+        CVERSION = True
+    else:
+        print("dynamic_programs_C_version built failed! Using python version!")
+        CVERSION = False
```
Comment (FrankLeeeee) [inline colossalai/fx/passes/algorithms/ckpt_solver_rotor.py:27]: This should be put inside the `solver_rotor` function.

Comment (Cypher30) [reply colossalai/fx/passes/algorithms/ckpt_solver_rotor.py:27]: okay

File: colossalai/fx/passes/algorithms/ckpt_solver_rotor.py [hunk 5]
```diff
@@ -323,34 +321,72 @@ def solver_rotor(gm: ColoGraphModule,
...
+    if not force_python:
+        try:
+            from .dynamic_programs_C_version import persistent_compute_table
+            CVERSION = True
+
+        # build module if module not found
+        except ModuleNotFoundError:
+            import subprocess
+            import os
+            print("dynamic_programs_C_version hasn't been built! Building library...")
```
Comment (FrankLeeeee) [inline colossalai/fx/passes/algorithms/ckpt_solver_rotor.py:351]: can replace print with colossalai logger.

Comment (Cypher30) [reply colossalai/fx/passes/algorithms/ckpt_solver_rotor.py:351]: okay

**Pull request merged**